primitives/phragmen/compact/src/lib.rs

```diff
@@ -0,0 +1,456 @@
+/// 	votes8: Vec<(AccountId, [(AccountId, W); 7usize], AccountId)>,
+/// 	votes9: Vec<(AccountId, [(AccountId, W); 8usize], AccountId)>,
+/// 	votes10: Vec<(AccountId, [(AccountId, W); 9usize], AccountId)>,
+/// 	votes11: Vec<(AccountId, [(AccountId, W); 10usize], AccountId)>,
+/// 	votes12: Vec<(AccountId, [(AccountId, W); 11usize], AccountId)>,
+/// 	votes13: Vec<(AccountId, [(AccountId, W); 12usize], AccountId)>,
+/// 	votes14: Vec<(AccountId, [(AccountId, W); 13usize], AccountId)>,
+/// 	votes15: Vec<(AccountId, [(AccountId, W); 14usize], AccountId)>,
+/// 	votes16: Vec<(AccountId, [(AccountId, W); 15usize], AccountId)>,
+/// }
+/// ```
+#[proc_macro]
+pub fn generate_compact_solution_type(item: TokenStream) -> TokenStream {
+	let CompactSolutionDef {
+		vis,
+		ident,
+		count,
+	} = syn::parse_macro_input!(item as CompactSolutionDef);
+
+	let account_type = GenericArgument::Type(Type::Verbatim(quote!(AccountId)));
+	let weight_type = GenericArgument::Type(Type::Verbatim(quote!(weight)));
+
+	let imports = imports().unwrap_or_else(|e| e.to_compile_error());
+
+	let compact_def = struct_def(
+		vis,
+		ident.clone(),
+		count,
+		account_type.clone(),
+		weight_type,
+	).unwrap_or_else(|e| e.to_compile_error());
+
+	let from_into_impl_assignment = convert_impl_for_assignment(
+		ident.clone(),
+		account_type.clone(),
+		count,
+	);
+
+	let from_into_impl_staked_assignment = convert_impl_for_staked_assignment(
+		ident,
+		account_type,
+		count,
+	);
+
+	quote!(
+		#imports
+		#compact_def
+		#from_into_impl_assignment
+		#from_into_impl_staked_assignment
+	).into()
+}
+
+fn struct_def(
+	vis: syn::Visibility,
+	ident: syn::Ident,
+	count: usize,
+	account_type: GenericArgument,
+	weight_type: GenericArgument,
+) -> Result<TokenStream2> {
+	if count <= 2 {
+		Err(syn::Error::new(
+			Span::call_site(),
+			"cannot build compact solution struct with capacity less than 2."
+		))?
+	}
+
+	let singles = {
+		let name = field_name_for(1);
+		quote!(#name: Vec<(#account_type, #account_type)>,)
+	};
+
+	let doubles = {
+		let name = field_name_for(2);
+		quote!(#name: Vec<(#account_type, (#account_type, #weight_type), #account_type)>,)
+	};
+
+	let rest = (3..=count).map(|c| {
+		let field_name = field_name_for(c);
+		let array_len = c - 1;
+		quote!(
+			#field_name: Vec<(
+				#account_type,
+				[(#account_type, #weight_type); #array_len],
+				#account_type
+			)>,
+		)
+	}).collect::<TokenStream2>();
+
+	let compact_def = quote! (
+		/// A struct to encode a `Vec<StakedAssignment>` or `Vec<_phragmen::Assignment>` of the phragmen module
+		/// in a compact way.
+		#[derive(Default, PartialEq, Eq, Clone, _sp_runtime::RuntimeDebug, _codec::Encode, _codec::Decode)]
+		#vis struct #ident<#account_type, #weight_type> {
+			#singles
+			#doubles
+			#rest
+		}
+	);
+
+	Ok(compact_def)
+}
+
+fn imports() -> Result<TokenStream2> {
+	let runtime_imports = match crate_name("sp-runtime") {
```

Comment:
No please don't do this.

The rule for macros is that you include one crate and this is the crate which re-exports the macro. In your case it would be `sp-phragmen`. When you require access to `sp_core` and `parity-scale-codec`, you will need to re-export them with `doc(hidden)` from `sp-phragmen`.

Reply:
fixed.